platform_plugin_aspects/signals.py

```diff
@@ -80,25 +81,27 @@ def on_user_profile_updated(instance):
     )
 
 
-def on_user_profile_updated_txn(**kwargs):
+def on_txn_wrapper(func, *args, **kwargs):
     """
-    Handle user_profile saves in the middle of a transaction.
+    Handle saves in the middle of a transaction.
 
     If this gets fired before the transaction commits, the task may try to
     query an id that doesn't exist yet and throw an error. This should postpone
     queuing the Celery task until after the transaction is committed.
     """
-    transaction.on_commit(
-        lambda: on_user_profile_updated(kwargs["instance"])
-    )  # pragma: no cover
+
+    def on_txn(**kwargs):
```

Comment:
I've confirmed that this doesn't work outside of dev mode. I'm not sure why, but am out of time to spend on it. I know this comes from trying to reduce the amount of boilerplate code, but haven't found a way to actually make it work. This pattern does work for me, and at least consolidates the functions to one place:

```
def on_user_profile_updated_txn(**kwargs):
    """
    Handle user_profile saves in the middle of a transaction.
    Handle saves in the middle of a transaction.
    If this gets fired before the transaction commits, the task may try to
    query an id that doesn't exist yet and throw an error. This should postpone
    queuing the Celery task until after the transaction is committed.
    """
    def on_user_profile_updated(instance):
        """
        Queues the UserProfile dump job when the parent transaction is committed.
        """
        # import here, because signal is registered at startup, but items in tasks are not yet able to be loaded
        from platform_plugin_aspects.tasks import (  # pylint: disable=import-outside-toplevel
            dump_data_to_clickhouse,
        )

        sink = UserProfileSink(None, None)
        dump_data_to_clickhouse.delay(
            sink_module=sink.__module__,
            sink_name=sink.__class__.__name__,
            object_id=str(instance.id),
        )

    transaction.on_commit(
        lambda: on_user_profile_updated(kwargs["instance"])
    )


# Connect the UserProfile.post_save signal handler only if we have a model to attach to.
# (prevents celery errors during tests)
_user_profile = get_model("user_profile")
if _user_profile:
    post_save.connect(
        on_user_profile_updated_txn, sender=_user_profile
    )  # pragma: no cover
```

Reply:
I will apply the changes